src/dataflow/src/render/upsert.rs

```diff
@@ -67,60 +80,22 @@ where
     // to specify that they believe that they have a large number of unique
     // keys, at which point materialize may be more performant if it runs
     // decoding/linear operators before deduplicating.
-
-    // This operator changes the timestamp from capability to message payload,
-    // and applies `as_of` frontier compaction. The compaction is important as
-    // downstream upsert preparation can compact away updates for the same keys
-    // at the same times, and by advancing times we make more of them the same.
-    let stream = stream.unary(Pipeline, "AppendTimestamp", move |_, _| {
-        let mut vector = Vec::new();
-        move |input, output| {
-            input.for_each(|cap, data| {
-                data.swap(&mut vector);
-                let mut time = cap.time().clone();
-                time.advance_by(as_of_frontier.borrow());
-                output
-                    .session(&cap)
-                    .give_iterator(vector.drain(..).map(|x| (x, time.clone())));
-            });
-        }
-    });
-
-    // Deduplicate records by key
-    let deduplicated = prepare_upsert_by_max_offset(&stream);
-
-    // Decode
-    let decoded = decode_upsert(
-        &deduplicated,
-        encoding,
-        key_encoding,
-        debug_name,
-        worker_index,
-    );
-
-    apply_linear_operators(&decoded, linear_operator, src_type)
-}
-
-/// Produces at most one entry for each `(key, time)` pair.
-///
-/// The incoming stream of `(key, (val, off), time)` records may have many
-/// entries with the same `key` and `time`. We are able to reduce this to
-/// at most one record for each pair, by retaining only the record with the
-/// greatest offset: its action summarizes the sequence of many actions that
-/// occur at the same moment and so are not distinguishable.
-fn prepare_upsert_by_max_offset<G>(
-    stream: &Stream<G, (SourceOutput<Vec<u8>, Vec<u8>>, Timestamp)>,
-) -> Stream<G, ((Vec<u8>, SourceData), Timestamp)>
-where
-    G: Scope<Timestamp = Timestamp>,
-{
-    stream.unary_frontier(
+    (&apply_as_of_frontier(stream, as_of_frontier)).unary_frontier(
```

Comment:
Does it make sense to keep the `apply_as_of_frontier` separate from the bulk of the logic here? Should we just push the use of `as_of_frontier` into the body of the logic below? Idk, but I would probably go and try it myself as a fast follow to this, unless there is a reason we don't do it (in which case, we should write it down).

Reply:
To me, the only reason to keep `apply_as_of_frontier` separate is that we have less visibility into the costs of `apply_as_of_frontier` if we fuse it in.

But if we have reasonable ability to reproduce performance issues on our own machines where we can profile things, I guess it's no longer necessary.

Reply:
Set this branch to automerge. Will create follow-on PR to fuse `apply_as_of_frontier` if you think it's worthwhile.